Zgamma plotting updates Pull Request

andersbarzdukas opened the pull request:
Hi All, 

I was looking at the merge conflicts between the zgamma_plotting_updates and master branch and found they were mostly minimal. 

Mainly there were a couple of duplicated functions, but beyond that there was mostly orthogonal changes to the files. 

I figured it would be easy to work from zgamma_plotting_updates and make any needed pushes before the merge while there were no current merge conflicts.

I may be jumping the gun, and we can revert to before the last push. The main change was the removal of pTt from the VH pTmiss category in the datacard and zg_utilities.cpp.

mcoshiro commented on src/zgamma/categorization_utilities.cpp:
```diff
@@ -0,0 +1,1487 @@
+
+namespace CatUtilities {
+
+  //Function checks if string 2 is in string 1. Used very briefly in making plots
+  bool contains(std::string string1, std::string string2){
+    return string1.find(string2,0)!=std::string::npos ? true : false;
+  }
```
Duplicate of Contains in `src/core/utilities.cpp`?

mcoshiro commented on src/zgamma/categorization_utilities.cpp:
```diff
@@ -0,0 +1,1487 @@
+  bool btag_DF_pass(TString year,float deep_flav,int lmt=1){
+    year.ReplaceAll("-","");
+    std::map<TString, std::vector<float>> btag_df_wpts{
+      {"2016", std::vector<float>({0.0614, 0.3093, 0.7221})},
+      {"2016APV", std::vector<float>({0.0614, 0.3093, 0.7221})},
+      {"2017", std::vector<float>({0.0521, 0.3033, 0.7489})},
+      {"2018", std::vector<float>({0.0494, 0.2770, 0.7264})},
+      {"2022", std::vector<float>({0.047,  0.245,  0.6734})},
+      {"2022EE", std::vector<float>({0.0499, 0.2605, 0.6915})},  
+      {"2023", std::vector<float>({0.0358, 0.1917, 0.6172})},
+      {"2023BPix", std::vector<float>({0.0359, 0.1919, 0.6133})}
```
These working points don't match the latest in nano2pico (out of date?). Also, we have some duplication with `src/zgamma/zg_utilities.cpp` L642-664. We should probably just decide on one place to keep the btag WPs in `draw_pico` and delete the other one (I don't have a preference as to which, but we'll have to fix the dependencies of whichever one we remove)

mcoshiro commented on src/zgamma/categorization_utilities.cpp:
```diff
@@ -0,0 +1,1487 @@
+    return L4_Idx(b)[0];
+  }
+
+  //Returns the difference in phi between to values
+  float get_dphi(float phi1, float phi2){
+     const double PI = 3.1415;
+     double dphi = fmod(fabs(phi2-phi1), 2.*PI);
+     dphi = dphi>PI ? 2.*PI-dphi : dphi;
+
+     return dphi;
+   }
```
Duplicate of `deltaPhi` in `src/core/utilities.cpp`?

mcoshiro commented on src/zgamma/controlregion_utilities.cpp:
```diff
@@ -0,0 +1,384 @@
+
+namespace CRUtilities{
+
+  bool contains(std::string string1, std::string string2){
+    return string1.find(string2,0)!=std::string::npos ? true : false;
+  }
```
Possible duplicate of Contains in `src/core/utilities.cpp`?

mcoshiro commented on src/zgamma/controlregion_utilities.cpp:
```diff
@@ -0,0 +1,384 @@
+  //This function adds all selections and reverses the selection at reverse
+  NamedFunc Nreverse1(std::vector<NamedFunc> vector_NamedFunc, unsigned int reverse){
+    NamedFunc return_NamedFunc = "1";
+    for(unsigned int idx = 0; idx<vector_NamedFunc.size(); idx++){
+      if(idx==reverse){return_NamedFunc = return_NamedFunc && !(vector_NamedFunc[idx]); continue;}  
+      return_NamedFunc = return_NamedFunc && vector_NamedFunc[idx];
+    }
+    return return_NamedFunc;
+  }
+
+  //Returns a NamedFunc replacing one selection (marked by skip)
+  NamedFunc Nreplace1(std::vector<NamedFunc> vector_NamedFunc, NamedFunc replace, unsigned int skip){
+    NamedFunc return_NamedFunc = "1";//vector_NamedFunc[start];
+    for(unsigned int idx = 0; idx<vector_NamedFunc.size(); idx++){
+      if(idx==skip){return_NamedFunc = return_NamedFunc && replace; continue;}  
+      return_NamedFunc = return_NamedFunc && vector_NamedFunc[idx];
+    }
+    return return_NamedFunc;
+  }
+
+  //Returns a NamedFunc with all but one selection (marked by skip)
+  NamedFunc Nminus1(std::vector<NamedFunc> vector_NamedFunc, unsigned int skip){
+//    NamedFunc return_NamedFunc = "1";
+//    unsigned int start = skip!=0 ?  0 : 1;
+    NamedFunc return_NamedFunc = "1";//vector_NamedFunc[start];
+    for(unsigned int idx = 0; idx<vector_NamedFunc.size(); idx++){
+      if(idx==skip){continue;}  
+      return_NamedFunc = return_NamedFunc && vector_NamedFunc[idx];
+    }
+    return return_NamedFunc;
+  }
+
+
+  //Returns a NamedFunc without all selections in the vector skip
+  NamedFunc Nminusk(std::vector<NamedFunc> vector_NamedFunc, std::vector<unsigned int> skip){
+    unsigned int idx_s = 0;
+    NamedFunc return_NamedFunc = "1";
+    for(unsigned int idx = 0; idx<vector_NamedFunc.size(); idx++){
+      if(idx_s < skip.size() && idx==skip[idx_s]){idx_s++; continue;}  
+      return_NamedFunc = return_NamedFunc && vector_NamedFunc[idx];
```
Not a suggestion, but just wanted to say I'm a fan of these functions (perhaps we can even move them to `src/core/named_func_utilities.cpp` someday?)

mcoshiro replied:
Oh, I see they are also in `src/zgamma/zg_functions.cpp`, so perhaps this copy can be removed?

andersbarzdukas replied:
I will move these to the named_func_utilities.cpp file and fix where they appear in any files.

mcoshiro commented on src/zgamma/zg_utilities.cpp:
```diff
@@ -128,6 +130,54 @@ namespace ZgUtilities {
   }
 
+  double AssignL1Error(const Baby &b) {
+    std::cout << " CRASH LEP" << std::endl;
```
debug message?

jdgrassi replied:
it's for when the crash is caused by the susy partner of LEP

mcoshiro replied:
H→Zγ TChiSlepSlep interpretation when? =p

mcoshiro commented on txt/colors_zgamma.txt:
```diff
@@ -1,13 +1,24 @@
+ dyjets   255 169  14
+ tt       231  99   0
+ vbszg    144 218 221
+ ttg      189  31   1
```
At some point, I think we decided to use this color for htozg and (#a96b59=169 107 89) for ttg

andersbarzdukas commented on src/zgamma/categorization_utilities.cpp:
```diff
@@ -0,0 +1,1487 @@
+  const NamedFunc baseline_nomll_VBF     = "dijet_m > 600";
+
+
+  const std::vector<std::vector<NamedFunc>> categories_selections_vector = {{max_Imini < 0.1, "ll_m[0] > 85 && ll_m[0] < 95"}, 
+                                                                            {"ll_m[0] > 85 && ll_m[0] < 95"},
+                                                                            {pTlly_mlly > 0.4, "ll_m[0] > 85 && ll_m[0] < 95"}, 
+                                                                            {max_Imini < 0.15, pTlly_mlly > 0.3, "ll_m[0] > 85 && ll_m[0] < 95", "met > 30"},
+                                                                            {"dijet_m > 600"}};
+
+  const std::vector<std::vector<NamedFunc>> categories_refit_selections_vector = {{max_Imini < 0.1, "ll_refit_m > 85 && ll_refit_m < 95"}, 
+                                                                                  {"ll_refit_m > 85 && ll_refit_m < 95"},
+                                                                                  {pTlly_mlly_refit > 0.4, "ll_refit_m > 85 && ll_refit_m < 95"}, 
+                                                                                  {max_Imini < 0.15, pTlly_mlly_refit > 0.3, "ll_refit_m > 85 && ll_refit_m < 95", "met > 30"},
+                                                                                  {"dijet_m > 600"}};
```
I agree these could probably be removed. The purpose of these functions is mainly if we want N-1 plots.

mcoshiro replied:
Ah, I was just suggesting to replace the redefinitions of the NamedFuncs in the vectors with the versions already defined above, not to remove the vectors, but I don't think it matters too much

andersbarzdukas commented:
> > Overall, things look good. I've made just a few cleanup suggestions, but I would say they are not necessary before merge
> 
> Hey Michael,
> 
> I will make the changes following the suggestions you have above before the push. I dont expect it to take too long, and I would like to do it now before I forget. I will mark those as resolved after pushing the changes.

Ok, I think I have addressed each point with the set of commits I just made. I checked that the draw_pico would compile and it does. 

I will wait for other comments before merging. If no comments are made by the weekly meeting time I will bring this up then.

Also as a bit of a heads-up, after the merge I will probably make a couple of pushes of plotting files for the plots that I have added to the AN.

andersbarzdukas commented on src/zgamma/categorization_utilities.cpp:
```diff
@@ -0,0 +1,1452 @@
+    if(b.nllphoton()>0){
+      return b.llphoton_costheta() -> at(0);
+    }
+    return ZgUtilities::cos_theta(b);
```
This pulls the cos_theta value from picos if it is there and if it is not the value is computed at runtime. These functions could be moved to the control region utilities, but they allow the creation of the plots for each category in the control regions where some of these values are not set in picos. 

The bool flag is set to false by default, so it will return the reco value.

andersbarzdukas commented on src/zgamma/categorization_utilities.cpp:
```diff
@@ -0,0 +1,1452 @@
+  }
+
+
+  std::vector<std::vector<float>> plotting_bins(bool plotting_run3){
```
This function probably could be removed? It mostly serves to create wide enough bins for the low stats categories.

andersbarzdukas commented:
Hi Joseph and Michael, there are still a few minor changes, but I think I will merge today and make some of those minor changes and add some an plotting files. If I dont hear back by end of day I will merge (will reach out on mattermost).
